public/app/core/components/SharedPreferences/SharedPreferences.tsx

```diff
@@ -67,12 +67,13 @@ export class SharedPreferences extends PureComponent<Props, State> {
       queryHistory: { homeTab: '' },
     };
 
-    this.themeOptions = [
-      { value: '', label: t('shared-preferences.theme.default-label', 'Default') },
-      { value: 'dark', label: t('shared-preferences.theme.dark-label', 'Dark') },
-      { value: 'light', label: t('shared-preferences.theme.light-label', 'Light') },
-      { value: 'system', label: t('shared-preferences.theme.system-label', 'System') },
-    ];
+    this.themeOptions = getBuiltInThemes(config.featureToggles.extraThemes).map((theme) => ({
+      value: theme.id,
+      label: t(`shared-preferences.theme.${theme.id}-label`, theme.name),
```

Comment:
Calls to `t()` must not contain expressions as we extract these statically. The solution for cases we have is to do something like this in grafana core:

```typescript
function getThemeTranslation(id: string) {
    switch (id) {
        case "dark": return t("shared.preferences.theme.dark-label", "Dark");
        case "light": return t("shared.preferences.theme.light-label", "Light");
        case "midnight-blue": return t("shared.preferences.theme.midnight-blue-label", "Midnight blue");
        default: id;
    }
}
```


(yes, we should have PR automation to automatically flag this for you ✨)

Reply:
@joshhunt but I can't duplicate this list another time, it's already duplicate in backend and frontend. I can't duplicate it here a third time. Can we move the t function to grafana/data so I can use it from there?

Reply:
The whole point of the registry (and we have many like this) is to not have to duplicate a bunch of names and ids

Reply:
I just can't accept your proposal as a workable solution for this, this place and the registry are going to get out of sync right away

Reply:
We've been trying pretty hard to avoid translating grafana/data:

 - We don't want grafana/data to depend on our translation framework (i18next)
 - It's a [bit of work/overhead to set up i18next](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/utils/i18n.tsx) (but not impossible) there also to let consumers not worry about it's existance
 - Creates more of a weird dependency between the grafana instance runtime and the 'data' library.

Reply:
Trying to think of the workaround/middleground? 

Must the registry exist in @grafana/data? Can it exist in core, or in @grafana/ui?

Reply:
I think a solution that _could_ work is to pass in `t` as an argument into the registry function. This would allow consumer (Grafana) to own the UI/translation, and initialisation of translations.

```typescreipt
const themeRegistry = new Registry<ThemeRegistryItem>((t: TranslationFunc) => {
  return [
    { id: 'system', name: t("shared.preferences.theme.system-label", "System"), build: getSystemPreferenceTheme },
    { id: 'dark', name: t("shared.preferences.theme.dark-label", "Dark"), build: () => createTheme({ colors: { mode: 'dark' } }) },
    ...etc
  ];
});
```

This works (I think) because the i18n extract is 'dumb' in that it just looks for function called `t()` regardless of where it's from.

Unsure how i feel about how 'ideal' this is, but it's another approach...

Reply:
@joshhunt hmm.. wanted to keep it close the core default themes and the palette files, but think it's not that important, can move it to the app, but no... it's needed from grafana/runtime config, hm... argh! 

I could make it so that config.theme2 is just the default theme, and I reset it later when the app boots and hope no one read it before then